### PR TITLE
Update ai to check for game end condition on Resume

### PR DESCRIPTION
**Pull request by bofstein**
Love this game!  Found an edge case while playing a lot where, since board.js checks for game end condition (no valid sets on the table, no cards left in deck) before it drew 3 cards, if the final draw left the board with no sets, the game could never end. Moving this check to the ai Resume function makes it check after cards are drawn instead, which should catch all instance instead of missing that 3% edge case, since the ai resumes at the start of each play.

I tested the change by speeding up the ai solve speed and animation speed and playing until I hit that state and confirming it ended the game now. You can tell it succeeded here because the bot has 23 sets: 81 cards - 12 on the table without any sets = 69 cards = 23 sets; this is where in the current version it would just be stuck since there were no sets and no cards to add now, but when it last checked there were 3 cards left.

image

### CODE REVIEW

**Comment by hugosaintemarie** [comment]
Hi @bofstein, and thank you very much for this PR!

I haven't worked on this project in almost three years so I don't remember how it all works to be honest. But, I'm going to trust that you've tested it and feel confident that moving that check (`numValidSets === 0 && deck.stock.length === 0`) to `ai.resume()` didn't break anything else 😁 

Thanks again for pointing out and fixing this edge case! Appreciate you making contributions to this old side project of mine 🙂

File: js/ai.js [hunk 1]
```diff
@@ -67,6 +67,14 @@ export default {
         this.test = 0;
         this.foundSet = false;
 
+        // have ai check each time it resumes if game should end; this logic used to be in board.js before card draw
+        let numValidSets = this.findValidSets().length
+        if (isDev)  console.log(`ai sets found: ${numValidSets}, cards left: ${deck.stock.length`);
```
Comment (zachkirsch) [inline js/ai.js:72]: This is a syntax error - missing the closing `}` at the end of the line. i.e. should be ``.length}`);``

Comment (hugosaintemarie) [reply js/ai.js:72]: Good catch @zachkirsch! Thanks. Just fixed on master. :)

**Comment by bofstein** [comment]
> Hi @bofstein, and thank you very much for this PR!
> 
> I haven't worked on this project in almost three years so I don't remember how it all works to be honest. But, I'm going to trust that you've tested it and feel confident that moving that check (`numValidSets === 0 && deck.stock.length === 0`) to `ai.resume()` didn't break anything else 😁
> 
> Thanks again for pointing out and fixing this edge case! Appreciate you making contributions to this old side project of mine 🙂

Thank you for merging!  This was a fun learning experience for me and I appreciate the ability to contribute.  I am not actually a programmer (aside from some complex Google Sheets formulas in Advent of Code or doing statistical analysis in R) - this is the first thing I've ever done in java script, I hope your trust is not misplaced.  I did test that it worked for me locally, fixed the bug, and didn't seem to mess anything up.

I noticed the live version at https://playset.netlify.app/ seems to still have the same bug, as it was stuck here at 23 sets made and wouldn't end since there were no cards left nor sets on table (unless I missed any).  I'm curious is that a separate deploy and this change isn't out on that version, or if it is out and the change didn't actually fix it for some reason.
image
image

File: js/ai.js (hunk 1 above)
Comment (bofstein) [reply js/ai.js:72]: Thanks for catching that!  I had it correct on my local copy, not sure how I messed that up pasting over to github.  Much appreciated, I'll fix my own branch too

**Comment by hugosaintemarie** [comment]
Good catch @bofstein! A couple thing were wrong with the configuration of the deployment on Netlify. This should now be fixed! If the changes are not reflected this time, make sure you clear your cache.

Let me know if you find any other issue 🙂 Have fun playing!

**Comment by bofstein** [comment]
It appears to be working now!  When I first opened it was black screen for about a few minutes even after a hard refresh so I was worried I did break it, but then it loaded and has worked fine since, and I eventually got to the case (23 sets found) to confirm it now ends properly!
image

I did find one more even smaller bug while testing: if you make an error before you find your first set, you don't get negative points for it (since you didn't have any points yet to subtract from).

I may investigate that next for fun/learning, though it could also be considered just a helpful boon for those starting out you don't want to nerf :)

**Comment by hugosaintemarie** [comment]
> I did find one more even smaller bug while testing: if you make an error before you find your first set, you don't get negative points for it (since you didn't have any points yet to subtract from).

I'm not 100% sure anymore, but I believe that in the physical version of the game, you count points by stacking valid sets you found in front of you, and when you make a mistake, you toss one of these away. With no valid sets found, a mistake doesn't subtract anything. :)

**Comment by bofstein** [comment]
Ooh great point! I just checked my copy and it says the player "loses one point" for making an error, and at the end is "given" one point per Set. I always read that as a negative point that are added to the positives at the end, but I could see the other way too, that you can't "lose" a point you don't have yet. Your way is probably how more people actually play.
PXL_20231030_144124016
